### PR TITLE
add Stack.ts

### PR DESCRIPTION
**Pull request by fredzqm**
Queue.ts is a FIFO throttler. I added Stack.ts, which is a FILO throttler.

Using DFS should significantly reduce the memory footprint of large delete CLI.

### CODE REVIEW

File: src/test/throttler/throttler.spec.ts [hunk 1]
```diff
@@ -3,14 +3,20 @@ import * as sinon from "sinon";
...
 const { expect } = chai;
 
-import Queue from "../queue";
+import Queue from "../../throttler/queue";
+import Stack from "../../throttler/stack";
 
 const TEST_ERROR = new Error("foobar");
 
-describe("Queue", () => {
+const throttlerTest = (Throttler: any) => {
```
Comment (fredzqm) [inline src/test/throttler/throttler.spec.ts]: Any good suggestion for `Throttler`'s type?

File: src/test/throttler/queue.spec.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  it("should have default name of queue", () => {
+    const queue = new Queue({});
+    expect(queue.name).to.equal("queue");
+  });
+
+  it("should be first-in-first-out", () => {
+    const queue = new Queue({});
+    expect(queue.hasWaitingTask()).to.equal(false);
+
+    queue.total++;
```
Comment (bkendall) [inline src/test/throttler/queue.spec.ts]: _huge_ -1 on this behavior

Use the queue like you _actually_ would to test it. Don't set internal state like this and write assertions against it.

File: src/test/throttler/stack.spec.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  it("should have default name of stack", () => {
+    const stack = new Stack({});
+    expect(stack.name).to.equal("stack");
+  });
+
+  it("should be first-in-last-out", () => {
+    const stack = new Stack({});
+    expect(stack.hasWaitingTask()).to.equal(false);
+
+    stack.total++;
```
Comment (bkendall) [inline src/test/throttler/stack.spec.ts]: Again, use the class like you _actually_ would to test it. Don't just set internal state.

File: src/throttler/throttler.ts [hunk 4]
```diff
@@ -75,6 +74,18 @@ export class Queue<T> {
...
   }
 
+  /**
+   *
+   * @return {boolean} true if there are unscheduled task waiting to be scheduled.
+   */
+  public abstract hasWaitingTask(): boolean;
+
+  /**
+   *
```
Comment (bkendall) [inline src/throttler/throttler.ts]: Remove the extra empty comment line (that contains only a single `*`) - same above

File: src/test/throttler/throttler.spec.ts (hunk 1 above)
Comment (bkendall) [reply src/test/throttler/throttler.spec.ts]: `Stack | Queue`?

Comment (fredzqm) [reply src/test/throttler/throttler.spec.ts]: No, it won't work because `Throttler` is not an instance of Stack or Queue. It's the classOf them.

Comment (bkendall) [reply src/test/throttler/throttler.spec.ts]: ah! you'll want to see this part of the documentation: https://www.typescriptlang.org/docs/handbook/interfaces.html and scroll down to "Class Types"

**Comment by fredzqm** [comment]
@bkendall Could you take another look? Thx

File: src/test/throttler/throttler.spec.ts [hunk 5]
```diff
@@ -216,4 +227,90 @@ describe("Queue", () => {
...
+  /**
+   * A promise that wait until this task starts executing.
+   */
+  startExecutePromise: Promise<any>;
+}
+
+export const createTask = (name: string, resolved: boolean) => {
+  return new Promise<Task>((res) => {
+    let resolve: (value?: any) => void = () => {
+      throw new Error();
```
[the remark this answers is not in the document]
Comment (bkendall) [reply src/test/throttler/throttler.spec.ts]: So the idea is: if one of these _aren't_ overridden by your logic below that overrides them, you want the test to throw an error - and if it does, the error message should help you figure out what went wrong.

File: src/throttler/throttler.ts [hunk 6]
```diff
@@ -75,6 +74,16 @@ export class Queue<T> {
     }
   }
 
+  /**
+   * @return {boolean} true if there are unscheduled task waiting to be scheduled.
```
Comment (bkendall) [inline src/throttler/throttler.ts]: I missed this: remove the type - Typescript already knows the type from the signature.

```suggestion
   * @return `true` if there are unscheduled task waiting to be scheduled.
```

File: src/throttler/throttler.ts [hunk 7]
```diff
@@ -75,6 +74,16 @@ export class Queue<T> {
...
+   */
+  public abstract hasWaitingTask(): boolean;
+
+  /**
+   * @return {number} the index of the next task to schedule.
```
Comment (bkendall) [inline src/throttler/throttler.ts]:
```suggestion
   * @return the index of the next task to schedule.
```

**Pull request merged**